backend/logic/users/users.js

```diff
@@ -181,20 +205,117 @@ var getStudentData = async studentID => {
 
   let data;
   let conn = await mysql.getNewConnection();
-  let courses, major, minor;
+  let completedCourses, incompletedCourses, major, minors;
   try {
-    courses = await conn.query(
+    completedCourses = await conn.query(
       `SELECT course_code, semester
     FROM course_offerings
     WHERE (id, semester)
     IN (SELECT offering_id, semester FROM student_course_offerings WHERE student_id = ?);`,
       [studentID]
     );
-    major = await conn.query(`SELECT curriculum_name FROM student_majors WHERE student_id = ?;`, [studentID]);
-    minor = await conn.query(`SELECT curriculum_name FROM student_minors WHERE student_id = ?;`, [studentID]);
+    major = await conn.query(
+      `SELECT curriculum_name FROM student_majors WHERE student_id = ?;`,
+      [studentID]
+    );
+    minors = await conn.query(
+      `SELECT curriculum_name FROM student_minors WHERE student_id = ?;`,
+      [studentID]
+    );
+
+    let currYear = new Date().getFullYear();
+    let currMonth = new Date().getMonth();
+    let fallSem, winterSem;
+    let curriculumName = major[0].curriculum_name;
+
+    if (currMonth < 3) {
+      fallSem = "";
+      winterSem = "W" + currYear;
+    } else {
+      fallSem = "F" + currYear;
+      winterSem = "W" + (currYear + 1);
+    }
+
+    let incompleteCore = await conn.query(
+      `SELECT course_code, semester 
+      FROM course_offerings 
+      WHERE (course_code 
+              NOT IN (SELECT course_code 
+                      FROM course_offerings 
+                      WHERE (id, semester) 
+                      IN (SELECT offering_id, semester 
+                          FROM student_course_offerings 
+                          WHERE student_id = ?))) 
+      AND (semester = ? OR semester = ?) 
+      AND (course_code 
+          IN (SELECT course_code 
+              FROM curriculum_core_classes 
+              WHERE curriculum_name = ?));`,
+      [studentID, fallSem, winterSem, curriculumName]
+    );
+    let desiredTC = await conn.query(
+      `SELECT course_code, semester 
+      FROM course_offerings 
+      WHERE (course_code 
+            NOT IN (SELECT course_code 
+                    FROM course_offerings 
+                    WHERE (id, semester) 
+                    IN (SELECT offering_id, semester 
+                        FROM student_course_offerings 
+                        WHERE student_id = ?))) 
+      AND (semester = ? OR semester = ?) 
+      AND (course_code 
+          IN (SELECT course_code 
+              FROM curriculum_tech_comps 
+              WHERE curriculum_name = ?)) 
+      AND (course_code
+          IN (SELECT course_code 
+              FROM student_desired_courses 
+              WHERE student_id = ?));`,
+      [studentID, fallSem, winterSem, curriculumName, studentID]
```

Comment:
explain the query to me to help me understand it

Reply:
Select all courses that are not in the set of completed courses AND that are in the correct semester (offered next year) AND that are in the set of valid techcomps for the major AND that are in the set of desired courses to take

Reply:
Is there a way to reduce the number of tables being queried? Or can you pull data from the tables you need and put them in a hashtable? As more data gets inserted in the database, these queries are going to become way too slow

Reply:
Reduced it by one. I know it's still not optimal but I suggest we just make a low priority task for further optimizing if that's fine given there are more urgent tasks for now.